Documentation/design-docs/first-class-structs.md

```diff
@@ -15,6 +15,261 @@ Primary Objectives
+encountered by most phases of the JIT:
+* `GT_INDEX`: This is transformed to a `GT_IND`
+* `GT_FIELD`: This is transformed to a `GT_LCL_FLD` or a `GT_IND`
+* `GT_MKREFANY`: This produces a "known" struct type, which is currently obtained by
+  calling `impGetRefAnyClass()` which is a call over the JIT/EE interface. This node is always
+  eliminated, and its source address used to create a copy. If it is on the rhs
+  of an assignment, it will be eliminated during the importer. If it is a call argument it will
+  be eliminated during morph.
+  * The presence of any of these in a method disables struct promotion. See `case CEE_MKREFANY` in the
+    `Importer`, where it is asserted that these are rare, and therefore not worth the trouble to handle.
+
+### Struct “objects” as lvalues
+
+* The lhs of a struct assignment is a block node or local
+  * `GT_OBJ` nodes represent the “shape” info via a struct handle, along with the GC info
+    (location and type of GC references within the struct).
+    * These are used only to represent struct values that contain GC references.
+  * `GT_BLK` nodes represent struct types with no GC references, or opaque blocks of fixed size.
+    * These have no struct handle, resulting in some pessimization or even incorrect
+      code when the appropriate struct handle can't be determined.
+    * These never represent lvalues of structs that contain GC references.
+    * Proposed: Like `GT_OBJ`, these would have an index into the `ClassLayout` cache.
+  * `GT_STORE_OBJ` and `GT_STORE_BLK` have the same structure as `GT_OBJ` and `GT_BLK`, respectively
+    * `Data()` is op2
+  * `GT_DYN_BLK` and `GT_STORE_DYN_BLK` (GenTreeDynBlk extends GenTreeBlk)
+    * Additional child `gtDynamicSize`
+    * Note that these aren't really struct types; they represent dynamically sized blocks
+      of arbitrary data.
+  * For `GT_LCL_VAR` nodes, the shape information is obtained via the `LclVarDsc`
+
+### Struct “objects” as rvalues
+
+Structs only appear as rvalues in the following contexts:
+
+* On the RHS of an assignment
+  * The lhs provides the “shape” for an assignment. Note, however, that the LHS isn't always available
+    to optimizations, which has led to pessimization, e.g. 
+    [#23739 Block the hoisting of TYP_STRUCT rvalues in loop hoisting](https://github.com/dotnet/coreclr/pull/23739)
+
+* As a call argument
+  * In this context, it must be one of: `GT_OBJ`, `GT_LCL_VAR`, `GT_LCL_FLD` or `GT_FIELD_LIST`
+
+* As an operand to a hardware or SIMD intrinsic (for `TYP_SIMD*` only)
+  * In this case the struct handle is generally assumed to be unneeded, as it is captured (directly or
+    indirectly) in the `GT_SIMD` or `GT_HWIntrinsic` node.
+
+After morph, a struct-typed value on the RHS of assignment is one of:
+* `GT_IND`: in this case the LHS is expected to provide the struct handle
+  * Proposed: `GT_IND` would no longer be used for struct types
+* `GT_CALL`
+* `GT_RET_EXPR`
```

Comment:
Hmm, `GT_RET_EXPR` after morph?

Reply:
Yes, and `GTK_NOTMIR` if we had such a thing. Lacking that we should probably assert that we don't see it in global morph, it's supposed to be gone after inlining.

Reply:
And apparently the case `GT_RET_EXPR` in `gtGetStructHandleIfPresent` isn't reached, though it's hard to be 100% sure as there are quite a few callers of `gtGetStructHandleIfPresent`.